zk/stages/stage_sequence_execute.go

```diff
@@ -47,6 +44,83 @@ func SpawnSequencingStage(
 		defer tx.Rollback()
 	}
 
+	lastBatch, err := stages.GetStageProgress(tx, stages.HighestSeenBatchNumber)
+	if err != nil {
+		return err
+	}
+
+	highestBatchInDS, err := cfg.datastreamServer.GetHighestBatchNumber()
+	if err != nil {
+		return err
+	}
+
+	if lastBatch >= highestBatchInDS {
+		err = sequencingStageStep(s, u, tx, ctx, cfg, quiet, nil)
+		if err != nil {
+			return err
+		}
+	} else {
+		log.Info(fmt.Sprintf("[%s] Last batch %d is lower than highest batch in datastream %d, resequencing...", s.LogPrefix(), lastBatch, highestBatchInDS))
+
+		latestForkId, err := stages.GetStageProgress(tx, stages.ForkId)
+		if err != nil {
+			return err
+		}
+		dsClient := client.NewClient(ctx, fmt.Sprintf("%s:%d", cfg.zk.DataStreamHost, cfg.zk.DataStreamPort), cfg.zk.DatastreamVersion, cfg.zk.L2DataStreamerTimeout, uint16(latestForkId))
+
```

Comment:
Do we need to call dsClient.Start(), before invoking reading batches?

P.s. this comment can be ignored if we use the "singleton" dsClient instance if I'm not mistaken.

Reply:
Good catch! We will probably need it :D